Try new voyage algorithm

paulbert opened the pull request:
So I had a sort of "Aha!" moment today about Voyage crew optimization.  Rather than just trying a bunch of combinations of crew and picking the best, we can pick the crew that provides the biggest boost to the voyage time at each step.  It has the added bonus of not over-using primary and secondary skills when the voyage will end before those skills start failing.

Basic idea of it:

for slot 1 to 12
  for each crew
    if can fit in available slot, calculate voyage time
  put crew with best time in open, valid slot

It checks for matching traits and also will go back to try different combinations if multiple slots were available for a crew.

Not sure that it's finding the 100% optimal choice right now, but it's finding better selections than the current release (haven't tried the multi-thread version).

It's running in JavaScript for me using a modified version of the @ChewableCPP [CodePen](https://codepen.io/somnivore/pen/Nabyzw) code for the voyage time calculation.

ChewableCPP commented:
This is interesting.

It's very much like earlier idea that I discounted because for various reasons (mainly that it was more complicated and I wasn't convinced it would be as good).

But this makes me think I may have dismissed it prematurely.

I'm working on an update to the native code that does something closer to this, but should (I hope) be both faster and optimal. I might have a PR for it within the next day.

We do want those toggles, I think, regardless of what we do about the voyage algorithm.

I haven't entirely understood the js voyage algorithm here so I'm not sure if it's "good enough". I can definitely see how it would give better results in at least some situations, though.

If we're right that brute force can essentially be avoided, a js solution could be fast enough to avoid needing native code. The only problem is that I'm much more comfortable in C++ than js.

One important thing to keep in mind about the js version here though is that the duration calculation is not as optimal as the current native version, because the native version takes into account the actual proficiency rolls directly, which the CodePen version cannot since it doesn't have access to that level of detail. So if we do decide to go with js, at least that ought to be remedied.

ChewableCPP commented:
I just pushed PR #13.

I realized why I gave up on my original idea earlier. It reduces to brute force for optimality. There are situations where swapping only works if you do it recursively, but then there is no obvious stopping point to the recursion. I can think of heuristics to terminate the recursion early, but then it starts to resemble what I've already done with presorting.

Anyway, what I did do is add an iterative swapping refinement pass on top of the brute force. This will help dramatically in the cases where the presorting ends up with too much crew of one skill (usually the primary). It's still not quite optimal. I can think of additional ways to improve this further to make it closer to optimal, but I think theoretically there may not be a 100% optimal solution given worst case roster scenarios that doesn't rely on brute force because of all the considerations on the duration calculation.

IAmPicard commented:
@paulbert , this looks like great stuff. Initially went the native / C++ route because I found JavaScript to be way too slow.

I was wondering if you can add a toggle to select which algorithm to go with? 😄 That way we can keep iterating on both, and let folks pick and choose what works best?

I'm considering adding an option to let people upload their (anonymized) inputs (crew stats, voyage slots config); once we have a large data set of those, we can run several versions of various algorithms to see which works best. Nut sure where I'd upload that stuff or if folks would be comfortable uploading their stats from the app.

IAmPicard commented on src/components/CrewRecommendations.js:
```diff
@@ -228,44 +239,49 @@ export class VoyageCrew extends React.Component {
+			traitScoreBoost: 200
 		};
 
+		if (this.state.includeFrozen === false) {
```
We probably want to move this filtering down into the algorithm implementation itself (in your case that would be the calculateBestVoyage function).

ChewableCPP commented:
I agree that we could just keep both in there.

But please update the duration calculation to match the native code with respect to the proficiency ranges. We don't want to mislead users into thinking that one or the other gives better results due to this difference. It should give the same estimate for the same result in both versions.

Data collection is interesting and may be worth doing, but there is a simpler thing we can do first:
* Run for every combination of pri/sec
* And every combination of the other options (e.g. frozen/adventuring)

And most important - the duration estimate output needs to match for these tests to be useful.

Another idea I was considering earlier is to solve this via "tournament". I didn't mention it because I'm hoping we can simply converge at single "ideal" solution, but in case we somehow can't, we could just run all variants and return the best result.

ChewableCPP commented:
Anther idea for testing - do a chaos test where random (seeded) crew is ignored, over a bunch of random seeds.

ChewableCPP commented:
I just rewrote the native algorithm to incorporate your idea somewhat. See https://github.com/IAmPicard/StarTrekTimelinesSpreadsheet/pull/13#issuecomment-370703455.

paulbert commented:
I like the idea to have the option to use either, so I added.  I called this one "fast" and the other "thorough," but can change that if you have a better idea.  Maybe just "Strategy 1" & "Strategy 2".

Updated the duration calculation to match the cpp.  If we want to keep both algorithms long term, we should probably expose just the voyage duration calculation from cpp to the app as another NativeExtension function so any changes to that are reflected in both.

Looking at just a few cases so far I'm seeing both algorithms produce a better time.

ChewableCPP commented:
Agreed that ideally there should be one copy of the duration calculation to maintain, not two.

I don't know what overhead there might be for the binding.
